model/versioning/Versioned.php

```diff
@@ -944,13 +944,113 @@ public function findOwned($recursive = true, $list = null)
 	 * @param ArrayList $list Optional list to add items to
 	 * @return ArrayList list of objects
 	 */
-	public function findOwners($recursive = true, $list = null)
-	{
-		// Find objects in these relationships
-		return $this->findRelatedObjects('owned_by', $recursive, $list);
+	public function findOwners($recursive = true, $list = null) {
+		if (!$list) {
+			$list = new ArrayList();
+		}
+
+		// Build reverse lookup for ownership
+		// @todo - Cache this more intelligently
+		$rules = $this->lookupReverseOwners();
+
+		// Hand off to recursive method
+		return $this->findOwnersRecursive($recursive, $list, $rules);
+	}
+
+	/**
+	 * Find objects which own this object.
+	 * Note that objects will only be searched in the same stage as the given record.
+	 *
+	 * @param bool $recursive True if recursive
+	 * @param ArrayList $list List to add items to
+	 * @param array $lookup List of reverse lookup rules for owned objects
+	 * @return ArrayList list of objects
+	 */
+	public function findOwnersRecursive($recursive, $list, $lookup) {
+		// First pass: find objects that are explicitly owned_by (e.g. custom relationships)
+		$owners = $this->findRelatedObjects('owned_by', false);
```

Comment:
In your example, it would probably calculate the ownership twice, and the list would be the same set.

It does rely on `$object->Owned()->Owner() === $object` though, and if user code is doing custom relations that violate this, then they probably should be using custom relationships on both ends.